### DISCUSSION TITLE
Error handling on coerce

### DISCUSSION THREAD

**Discussion by Paddidox** [colinhacks/zod#1800] [Q&A] [answered]
I believe to have found a non critical bug, but I am not sure as I bring only limited knowledge in JS.

Code:

```ts
let array = ["1", "string"];

let validate = z.coerce
                .number()
                .array()
                .optional()
                .safeParse(array);

console.log(validate.error.issues)
```

Result:
```ts
[
  {
    code: 'invalid_type',
    expected: 'number',
    received: 'nan',
    path: [ 1 ],
    message: 'Expected number, received nan'
  }
]
```

I'd have assumed to get "received: 'string'", got 'nan' instead.

**Comment by JacobWeisenburger** [comment, marked as answer]
Essentially what is going on is `z.coerce.number()` takes what is passed to `parse/safeParse` and passes it to `Number()` and then passes the result to `z.number()`.

```ts
const result = z.coerce.number().safeParse( 'string' )
!result.success && console.log( result.error.issues[ 0 ].message )
// Expected number, received nan

console.log( Number( 'string' ) )
// NaN
```

Let me know if you have any other questions.

**Reply by oskstr** [reply]
Great addition with `coerce`! But I certainly ran into some gotchas when trying to use this to parse query parameters.

E.g.
```ts
const params = z.object({
    "page-nr": z.coerce.number(),
});

params.parse({ "page-nr": "" });    // { "page-nr": 0 }
params.parse({ "page-nr": "foo" }); // Error: "Expected number, received nan"
```

Is there a better way to handle these cases cleanly? Would be nice to be able to return something like:
`"Validation error: Expected number, received \"foo\" at \"page-nr\""` to the caller.

**Reply by colemars** [reply]
Some additional weirdness here - if a field is both required and coerced the error message for failed coercion overrides the "Required" error message. This doesn't seem ideal.
